File: leetcode/Algorithms/23.MergeKSortedLists/Solution.java
import java.util.Comparator;
import java.util.PriorityQueue;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode(int x) { val = x; }
 * }
 */

public class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        Comparator<ListNode> comparator = new Comparator<ListNode>() {
            @Override
            public int compare(ListNode l1, ListNode l2) {
                return l1.val - l2.val;
            }
        };
        PriorityQueue<ListNode> queue = new PriorityQueue<>(comparator);
        for (ListNode list : lists)
            if (list != null) queue.add(list);

        ListNode dummyHead = new ListNode(0), cur = dummyHead;
        while (!queue.isEmpty()) {
            cur.next = queue.poll();
            cur = cur.next;
            if (cur.next != null) queue.add(cur.next);
        }

        return dummyHead.next;
    }
}

File: leetcode/Algorithms/23.MergeKSortedLists/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        length = len(lists)
        if length == 0:
            return None
        elif length == 1:
            return lists[0]
        else:
            return self.merge_two_list(self.mergeKLists(lists[:length / 2]), self.mergeKLists(lists[length / 2:]))

    def merge_two_list(self, l1, l2):
        dummy_head = cursor = ListNode(0)

        while l1 and l2:
            if l1.val < l2.val:
                cursor.next = l1
                l1 = l1.next
            else:
                cursor.next = l2
                l2 = l2.next
            cursor = cursor.next

        cursor.next = l1 or l2
        return dummy_head.next

